### DISCUSSION TITLE
macOS scrolling

### DISCUSSION THREAD

**Discussion by nico** [ghostty-org/ghostty#2524] [Feature Requests, Ideas] [answered]
After playing around a bit with Ghostty, several things related to scrolling are missing or feel a bit off:

- There's no overlay scrollbar (or any scrollbar at all) (#111)
- Fn-arrowleft doesn't jump to top of scrollback, fn-arrowup doesn't scroll up by a page (#2526)
- Scrolling feels weird and non-native -- is there a custom scroll curve?
  - repro: open terminal.app window with same font size and same win size
           as a ghostty window, put both on top of each other, run `ls -hl`
           to have ~700 lines of scrollback in both, then do a full
           two-finger fling in both windows. terminal.app clears the whole
           buffer, ghostty only gets about halfway through.

AppKit apps get all this for free when using an NSScrollView. I'm guessing since Ghostty renders using metal, using an NSScrollView isn't super convenient. But the suggestion from here is maybe workable: https://forums.developer.apple.com/forums/thread/651229

That is, have a dummy NSView as the NSScrollView's documentView and sync its size using a technique similar to https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/NSScrollViewGuide/Articles/SynchroScroll.html

Apologies if this has been tried before and doesn't work due to performance issues. I tried looking around a bit but didn't find a discussion about this.

**Comment by mitchellh** [comment, marked as answer]
> There's no overlay scrollbar (or any scrollbar at all)

#111 

> Fn-arrowleft doesn't jump to top of scrollback, fn-arrowup doesn't scroll up by a page

This is probably fixable today... depending on if AppKit sends some event to the macOS app. We learned we can't reliably detect `fn+<key>` at a low level but if AppKit sends a key equivalent or something for this then we can do it.

Ghostty already has binding actions to jump scroll to top or up a single page so its just a matter of triggering that.

> Scrolling feels weird and non-native -- is there a custom scroll curve?

Yes, besides Terminal.app I think every Terminal implements a custom scroll curve because we _don't_ get it for free (with the current raw NSView).

> Apologies if this has been tried before and doesn't work due to performance issues. I tried looking around a bit but didn't find a discussion about this.

It hasn't. We can't use a full NSScrollView as you already ascertained but I haven't looked into the more creative approaches to it, so let me read on that a bit. As you can see with #111, I absolutely want to support a scollbar in the GUI for both macOS and Linux so this is relevant.

**Reply by mitchellh** [reply]
> Fn-arrowleft doesn't jump to top of scrollback, fn-arrowup doesn't scroll up by a page

Found them:

https://developer.apple.com/documentation/appkit/nsstandardkeybindingresponding#3745907
`

**Reply by mitchellh** [reply]
I opened #2526

**Comment by mitchellh** [comment]
> AppKit apps get all this for free when using an NSScrollView. I'm guessing since Ghostty renders using metal, using an NSScrollView isn't super convenient. But the suggestion from here is maybe workable: https://forums.developer.apple.com/forums/thread/651229
> 
> That is, have a dummy NSView as the NSScrollView's documentView and sync its size using a technique similar to https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/NSScrollViewGuide/Articles/SynchroScroll.html

Upon further reading, I think this is a worthy approach to try. I'm going to post this over at #111 with some additional details.
